lib/ocpp/v201/smart_charging.cpp

```diff
@@ -224,4 +232,86 @@ std::vector<ChargingProfile> SmartChargingHandler::get_station_wide_tx_default_p
     return station_wide_tx_default_profiles;
 }
 
+bool SmartChargingHandler::is_overlapping_validity_period(int candidate_evse_id,
+                                                          ChargingProfile candidate_profile) const {
+    if (candidate_profile.chargingProfilePurpose == ChargingProfilePurposeEnum::TxProfile) {
+        // This only applies to non TxProfile types.
+        return false;
+    }
+
+    auto conflicts_with = [candidate_evse_id, &candidate_profile](
+                              const std::pair<int32_t, std::vector<ChargingProfile>>& existing_profiles) {
+        auto existing_evse_id = existing_profiles.first;
+        if (existing_evse_id == candidate_evse_id) {
+            return std::any_of(
+                existing_profiles.second.begin(), existing_profiles.second.end(),
+                [&candidate_profile](const ChargingProfile& existing_profile) {
+                    if (existing_profile.stackLevel == candidate_profile.stackLevel &&
+                        existing_profile.chargingProfileKind == candidate_profile.chargingProfileKind &&
+                        existing_profile.id != candidate_profile.id) {
+
+                        // Existing profile last forever should be updated to replace
+                        if (!existing_profile.validFrom.has_value() && !existing_profile.validTo.has_value()) {
+                            return true; // reject
+                        }
+
+                        // Cannot accept new profile that last forever when there is an existing profile
+                        if (!candidate_profile.validFrom.has_value() && !candidate_profile.validTo.has_value()) {
```

Comment:
If the existing profile has a `validTo` that ends before the candidate is applied (which is the moment it's received in this case), shouldn't the candidate be acceptable?